File: apps/cargo-tracker/src/shared/domain/model/location.ts
/**
 * 位置情報（共有カーネル）。UN/LOCODE で識別される港湾・地点。
 * 全コンテキストで共有する（domain-model Shared Domain）。
 */
const UNLOCODE_PATTERN = /^[A-Z]{5}$/;

export class Location {
  private constructor(readonly unlocode: string) {}

  static of(raw: string): Location {
    const normalized = raw.trim().toUpperCase();
    if (!UNLOCODE_PATTERN.test(normalized)) {
      throw new Error(`不正な UN/LOCODE: ${raw}（5 文字の英字コードが必要です）`);
    }
    return new Location(normalized);
  }

  sameAs(other: Location): boolean {
    return this.unlocode === other.unlocode;
  }
}

File: docs/article/source/go/apps/cargo-tracker/internal/shared/domain/location.go
package domain

import "errors"

// ErrInvalidLocation は UN/LOCODE が不正な場合に返される。
var ErrInvalidLocation = errors.New("location must be a 5-character UN/LOCODE")

// Location は UN/LOCODE で識別される港湾・地点を表す共有カーネルの値オブジェクト。
type Location struct {
	unLocode string
}

// NewLocation はバリデーション付きで Location を生成する（UN/LOCODE は 5 文字）。
func NewLocation(unLocode string) (Location, error) {
	if len(unLocode) != 5 {
		return Location{}, ErrInvalidLocation
	}
	return Location{unLocode: unLocode}, nil
}

// UnLocode は UN/LOCODE 文字列を返す。
func (l Location) UnLocode() string { return l.unLocode }

// SameAs は同一地点かどうかを返す。
func (l Location) SameAs(other Location) bool { return l.unLocode == other.unLocode }

// IsZero は未設定かどうかを返す。
func (l Location) IsZero() bool { return l.unLocode == "" }
